### PR TITLE
Added enumerations to represent Observable return types #280

### PR DESCRIPTION
**Pull request by antalszava**
**Context:**
The Observable class has an attribute `return_type`. So far the None and the "sample", "variance" and "expectation" String values were used for this attribute. Consequently, changing to a type-safe version was considered.

**Description of the Change:**
The PR creates an enumeration based class introducing the Sample, Variance and Expectation values. Includes moreover additional validation in the `Device.execute()` method for an unwanted `return_type`:
```
elif obs.return_type is not None:
    raise QuantumFunctionError("Unsupported return type specified for observable {}".format(obs.name))
```

**Benefits:**
Similar to the benefits of the solution in https://github.com/XanaduAI/pennylane/pull/277. One difference is that for the `return_type` the more generic Enum class is used as associating the values with integers is not needed.

**Possible Drawbacks:**
 n/a

**Related GitHub Issues:**
https://github.com/XanaduAI/pennylane/pull/277

### CODE REVIEW

**Comment by josh146** [comment]
Thanks for the PR @antalszava!

File: tests/test_operation.py [hunk 1]
```diff
@@ -418,6 +418,19 @@ class DummyOp(oo.Operation):
         with self.assertRaisesRegex(ValueError, "Must specify the wires"):
             DummyOp(0.54, 0, do_queue=False)
 
+    def test_observable_return_type_none(self):
+        """Test exception raised if no return type is specified"""
```
Comment (co9olguy) [inline tests/test_operation.py]: Does not appear to test for an exception, but rather that the returned value of the dummy observable is `None`

File: tests/test_device.py [hunk 2]
```diff
@@ -370,6 +372,47 @@ def test_sample_attribute_error(self, mock_device_with_paulis_and_methods):
...
+        # Make a observable without specifying a return operation upon measuring
+        obs = qml.PauliZ(0, do_queue=False)
+        obs.return_type = "SomeUnsupportedReturnType"
+        observables = [obs]
+
+        with pytest.raises(QuantumFunctionError, match="Unsupported return type specified for observable"):
+            mock_device_with_paulis_and_methods.execute(queue, observables)
+
+    def test_supported_observable_return_types(self, mock_device_with_paulis_and_methods):
+        """Check that an no error is raised if the return types of observables are supported"""
```
Comment (co9olguy) [inline tests/test_device.py]:
```suggestion
        """Check that no error is raised if the return types of observables are supported"""
```

File: tests/test_operation.py (hunk 1 above)
Comment (co9olguy) [inline tests/test_operation.py]: Does not appear to test for an exception, but rather that the returned value of the dummy observable is `None`

File: pennylane/measure.py [hunk 3]
```diff
@@ -61,7 +61,7 @@ def circuit(x, y):
 import pennylane as qml
 
 from .qnode import QNode, QuantumFunctionError
-from .operation import Observable
```
Comment (co9olguy) [inline pennylane/measure.py]: The CI service is picking up a cyclic import. Is it coming from this addition?

File: tests/test_device.py (hunk 2 above)
Comment (antalszava) [reply tests/test_device.py]: Fair point, thank you!

File: tests/test_operation.py (hunk 1 above)
Comment (antalszava) [reply tests/test_operation.py]: That is true! I've reworded the comment, thanks!

File: pennylane/measure.py (hunk 3 above)
Comment (josh146) [reply pennylane/measure.py]: This change shouldn't have any affect re:cyclic import, so it was likely there before.

File: pennylane/_device.py [hunk 4]
```diff
@@ -221,15 +222,16 @@ def execute(self, queue, observables, parameters={}):
             self.pre_measure()
 
             for obs in observables:
-                if obs.return_type == "expectation":
+                if obs.return_type == Expectation:
```
Comment (josh146) [inline pennylane/_device.py]: A minor comment, but it is usually best practice to compare enums by identity, see https://docs.python.org/3/library/enum.html#comparisons
```suggestion
                if obs.return_type is Expectation:
```

Comment (antalszava) [reply pennylane/_device.py]: Thanks for the valid point! I have changed it accordingly.

File: .github/CHANGELOG.md [hunk 5]
```diff
@@ -57,6 +57,10 @@
   they are calculated with Euler's formula.
   [#292](https://github.com/XanaduAI/pennylane/pull/292)
 
+* Creates an `ObservableReturnTypes` enumeration class introducing the Sample, 
+  Variance and Expectation. These new values can be assigned to the `return_type`
+  attribute of an `Observable`.
+  [#290](https://github.com/XanaduAI/pennylane/pull/290)
```
Comment (josh146) [inline .github/CHANGELOG.md]:
```suggestion
  [#290](https://github.com/XanaduAI/pennylane/pull/290)

```

**Pull request merged**